kolena/_experimental/classification/dataset.py

```diff
@@ -0,0 +1,258 @@
+# Copyright 2021-2025 Kolena Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+from typing import Any
+from typing import Dict
+from typing import Iterable
+from typing import Iterator
+from typing import List
+from typing import Optional
+from typing import Union
+
+import pandas as pd
+import tqdm
+
+from kolena import dataset
+from kolena._experimental.classification._matching import InferenceMatches
+from kolena._experimental.classification._matching import match_inferences
+from kolena.annotation import Label
+from kolena.annotation import ScoredLabel
+from kolena.dataset.dataset import _load_dataset_metadata
+from kolena.errors import IncorrectUsageError
+
+
+def merge(gt: Union[str, Label, ScoredLabel], inf: Union[str, Label, ScoredLabel]) -> Union[str, Dict[str, Any]]:
+    if isinstance(gt, str) and isinstance(inf, str):
+        return gt
+    if isinstance(gt, str):
+        return inf._to_dict()
+    if isinstance(inf, str):
+        return gt._to_dict()
+    return {**gt._to_dict(), **inf._to_dict()}
+
+
+def datapoint_metrics(
+    object_matches: InferenceMatches,
+) -> Dict[str, Any]:
+    tp = [merge(gt, inf) for gt, inf in object_matches.matched]
+    fp = object_matches.unmatched_inf
+    fn = object_matches.unmatched_gt
+    count_tp = len(tp)
+    count_fp = len(fp)
+    count_fn = len(fn)
+    precision = count_tp / (count_tp + count_fp) if count_tp + count_fp > 0 else 0
+    recall = count_tp / (count_tp + count_fn) if count_tp + count_fn > 0 else 0
+    f1 = 2 * precision * recall / (precision + recall) if precision + recall > 0 else 0
+    jaccard_index = count_tp / (count_tp + count_fp + count_fn) if count_tp + count_fp + count_fn > 0 else 0
+
+    return dict(
+        TP=tp,
+        FP=fp,
+        FN=fn,
+        count_TP=len(tp),
+        count_FP=len(fp),
+        count_FN=len(fn),
+        is_exact_match=len(fp) + len(fn) == 0,
+        Precision=precision,
+        Recall=recall,
+        F1_Score=f1,
+        Jaccard_Index=jaccard_index,
+    )
+
+
+def _iter_metrics(
+    pred_df: pd.DataFrame,
+    all_object_matches: List[InferenceMatches],
+    *,
+    batch_size: int = 10_000,
+) -> Iterator[pd.DataFrame]:
+    for i in tqdm.tqdm(range(0, pred_df.shape[0], batch_size)):
```

Comment:
I'm wondering if the creation of `all_object_matches` should be part of the tqdm progress bar. It seems like `datapoint_metrics()` is fast math and `match_inferences()` has sorting.